README.md

```diff
@@ -0,0 +1,71 @@
+# Minimal ROS 2 real-time cookbook recipes
+
+Collection of ROS 2 minimal examples showing how to leverage different POSIX and ROS 2 features to
+create real-time applications.
+
+
+* [minimal_memory_lock](minimal_memory_lock/README.md): shows how to lock the process memory and
+ pre-allocate dynamic memory to avoid memory page faults
+
+
+TODO: Add other packages after review (currently in rolling-experimental branch): 
+* minimal_deadline_qos: shows how to use the DDS deadline QoS policy
+* minimal_lifecycle: shows how to use node lifecycle to separate configuration and runtime
+ real-time phases
+* minimal_memory_check: shows how use verify that functions used in the real-time path
+does not allocate dynamic memory
+* minimal_memory_allocator: shows how to use different memory strategies to avoid dynamic
+ memory allocations during the runtime phase
+* minimal_loaned_messages: shows how to use loaned messages APIs
+* minimal_data_sharing: shows how to use different data sharing approached to avoid
+ blocking calls when sharing data between real-time and non real-time threads 
+* minimal_dds_tuning: shows how to use DDS specific configurations to tune real-time
+ related settings (TODO: add XML profiles for more DDS implementations)
+
+
+## Requirements
+
+- ROS 2 rolling release
+- Linux based operating system (RT_PREEMPT is not mandatory to run the examples)
+- Privileges to set priority, scheduling and memory lock limits. This can be achieved by having
+ root privileges or adding user privileges to `/etc/security/limits.conf`
+
+
+## Build instructions
+
+### Create workspace
+
+```bash
+$ mkdir -p ros2_realtime_examples_ws/src
+$ cd ros2_realtime_examples_ws/src
+$ git clone https://github.com/carlossvg/ros2-realtime-examples.git
+# Optional: Build rmw_iceoryx for zero-copy examples
+$ git clone https://github.com/ros2/rmw_iceoryx
+```
+
+### Install dependencies 
+
+Option 1: from debian packages:
+
+```bash
+$ sudo apt install ros-rolling-tlsf-cpp ros-rolling-tlsf \
+  ros-rolling-apex-test-tools ros-rolling-apex-containers ros-rolling-test-msgs -y
+```
+
+Option 2: Using [rosdep](http://wiki.ros.org/ROS/Tutorials/rosdep):
+
+```bash
+$ cd ros2_realtime_examples_ws
+$ rosdep install --from-paths src/ --ignore-src --rosdistro=rolling
```

Comment:
Two reasons:
- I was not sure if all the dependencies were available in rosdep. This is the case for rolling, so rosdep would be enough.
- In case you don't want to install rosdep. For example, we don't include rosdep in the RPI4 image. 

Since these are not strong reasons I can keep the rosdep option only to simplify.

Reply:
>I think we should figure out who the target audience for this code is. Right now it feels like code that proves that memory locking is working (almost like unit-tests). Maybe, we should target RT app developers instead who are probably more interested in a "sample app framework" for ROS? Something with all the scheduling and memory setup. We can still put some logging about page faults, as that could be the "sample application code", if that makes sense. This way you don't need any of the options. We can quickly document all the things in the README. I also have a blogpost coming that explains all these options in hopefully a week.

The idea is to provide minimal code snippets to the can easily integrated in other applications. With that in mind the application should be simple, easy to understand and showing one concept at each time. The "sample app framework" matches the description of the demo app. I would like to consolidate all the concepts we show in the minimal examples in this demo https://github.com/ros-realtime/community/issues/37. This is the reason I stopped working on the refactoring and started with the examples. I want to start with simple examples, refine our understanding, potentially create some helper functions, and finally consolidate all this work in a demo where we can show good real-time performance.